pennylane/io/qasm_interpreter.py

```diff
@@ -0,0 +1,327 @@
+"""
+
+import re
+from functools import partial
+from typing import Callable
+
+from pennylane import ops
+
+has_openqasm = True
+try:
+    from openqasm3.visitor import QASMNode, QASMVisitor
+except (ModuleNotFoundError, ImportError) as import_error:  # pragma: no cover
+    has_openqasm = False  # pragma: no cover
+
+NON_PARAMETERIZED_GATES = {
+    "ID": ops.Identity,
+    "H": ops.Hadamard,
+    "X": ops.PauliX,
+    "Y": ops.PauliY,
+    "Z": ops.PauliZ,
+    "S": ops.S,
+    "T": ops.T,
+    "SX": ops.SX,
+    "CX": ops.CNOT,
+    "CY": ops.CY,
+    "CZ": ops.CZ,
+    "CH": ops.CH,
+    "SWAP": ops.SWAP,
+    "CCX": ops.Toffoli,
+    "CSWAP": ops.CSWAP,
+}
+
+PARAMETERIZED_GATES = {
+    "RX": ops.RX,
+    "RY": ops.RY,
+    "RZ": ops.RZ,
+    "P": ops.PhaseShift,
+    "PHASE": ops.PhaseShift,
+    "U1": ops.U1,
+    "U2": ops.U2,
+    "U3": ops.U3,
+    "CP": ops.CPhase,
+    "CPHASE": ops.CPhase,
+    "CRX": ops.CRX,
+    "CRY": ops.CRY,
+    "CRZ": ops.CRZ,
+}
+
+
+class QasmInterpreter(QASMVisitor):
+    """
+    Overrides generic_visit(self, node: QASMNode, context: Optional[T]) which takes the
+    top level node of the AST as a parameter and recursively descends the AST, calling the
+    overriden visitor function on each node.
+    """
+
+    def __init__(self):
+        """
+        Checks that the openqasm3 package is available, otherwise raises an error.
+
+        Raises:
+            ImportError: if the openqasm3 package is not available.
+        """
+        if not has_openqasm:  # pragma: no cover
+            raise ImportError(
+                "QASM interpreter requires openqasm3 to be installed"
+            )  # pragma: no cover
+        super().__init__()
+
+    def visit(self, node: QASMNode, context: dict):
+        """
+        Visitor function is called on each node in the AST, which is traversed using recursive descent.
+        The purpose of this function is to pass each node to the appropriate handler.
+
+        Args:
+            node (QASMNode): the QASMNode to visit next.
+            context (dict): the current context populated with any locally available variables, etc.
+
+        Returns:
+            dict: The context updated after the compilation of the current node into Callables
+                to queue into a QNode.
+
+        Raises:
+            NameError: When a (so far) unsupported node type is encountered.
+        """
+        match node.__class__.__name__:
+            case "QubitDeclaration":
+                self.qubit_declaration(node, context)
+            case "ClassicalDeclaration":
+                self.classical_declaration(node, context)
+            case "QuantumGate":
+                self.quantum_gate(node, context)
+            # TODO: call appropriate handler methods here
+            case _:
+                raise NotImplementedError(
+                    f"An unsupported QASM instruction was encountered: {node.__class__.__name__}"
```

Comment:
Can we add the `handler_name = 'visit_' + node.__class__.__name__` etc logic here now instead of starting out with a switch statement? While we will still have the freedom to change our decisions as we progress, I think it's a good idea to start with as strong a foundation as we can in this first PR.